File: backend/internal/domain/property/amenities.go
package property

import "strings"

// CanonicalAmenities is the shared, canonical set of amenity codes. The web
// create form and the search filter both use this list (via GET /amenities) so
// stored data and filters always agree.
var CanonicalAmenities = []string{
	"wifi",
	"kitchen",
	"parking",
	"pool",
	"air-conditioning",
	"heating",
	"washer",
	"dryer",
	"tv",
	"workspace",
	"hot-tub",
	"gym",
	"pets-allowed",
	"elevator",
	"breakfast",
	"smoke-alarm",
}

var amenitySet = func() map[string]struct{} {
	m := make(map[string]struct{}, len(CanonicalAmenities))
	for _, a := range CanonicalAmenities {
		m[a] = struct{}{}
	}
	return m
}()

// CanonicalizeAmenity normalises a free-form amenity to a canonical code, or ""
// if it is not recognised.
func CanonicalizeAmenity(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	s = strings.ReplaceAll(s, "_", "-")
	s = strings.ReplaceAll(s, " ", "-")
	if _, ok := amenitySet[s]; ok {
		return s
	}
	return ""
}

// NormalizeAmenities maps inputs to canonical codes, dropping unknown values and
// duplicates while preserving order.
func NormalizeAmenities(in []string) []string {
	seen := make(map[string]struct{}, len(in))
	out := make([]string, 0, len(in))
	for _, a := range in {
		c := CanonicalizeAmenity(a)
		if c == "" {
			continue
		}
		if _, ok := seen[c]; ok {
			continue
		}
		seen[c] = struct{}{}
		out = append(out, c)
	}
	return out
}

File: web/src/amenities.js
// Fallback canonical amenity codes, mirroring the backend's
// property.CanonicalAmenities. The live list is fetched from GET /amenities;
// this is used as the initial value and a fallback if the request fails.
export const AMENITY_CODES = [
  'wifi',
  'kitchen',
  'parking',
  'pool',
  'air-conditioning',
  'heating',
  'washer',
  'dryer',
  'tv',
  'workspace',
  'hot-tub',
  'gym',
  'pets-allowed',
  'elevator',
  'breakfast',
  'smoke-alarm',
  // Accessibility codes (S161) — must stay in sync with backend
  // property.CanonicalAmenities so the EAA / ADA filter options render even
  // when GET /amenities fails and we fall back to this client-side list.
  'wheelchair-accessible',
  'step-free-entry',
  'wide-doorways',
  'roll-in-shower',
  'accessible-parking',
  'accessible-bathroom',
];
